aws/retry/middleware.go

```diff
@@ -0,0 +1,177 @@
+package retry
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	"github.com/awslabs/smithy-go/middleware"
+	"github.com/awslabs/smithy-go/transport/http"
+)
+
+type retryMetadata struct {
+	AttemptNum       int
+	AttemptTime      time.Time
+	MaxAttempts      int
+	AttemptClockSkew time.Duration
+}
+
+type retryMetadataKey struct{}
+
+type RetryMiddleware struct {
+	Retryer Retryer
+}
+
+func (r RetryMiddleware) ID() string {
+	return "Retry Middleware"
+}
+
+func (r RetryMiddleware) Name() string {
+	return r.ID()
+}
+
+func (r RetryMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (out middleware.FinalizeOutput, err error) {
+	const invocationIDHeader = "amz-sdk-invocation-id"
+
+	req := in.Request.(*http.Request)
+
+	// TODO: Don't see a compelling reason to make this a separate middleware that makes this available via context
+	invocationID, err := sdk.UUIDVersion4()
+	if err != nil {
```

Comment:
This might be helpful via logging, would make sense to be in a `Build` handler, separate from retry middleware implementation. Helpful if the retry middleware is replaced.